src/java/org/apache/cassandra/utils/JVMStabilityInspector.java

```diff
@@ -186,17 +205,39 @@ public static void userFunctionTimeout(Throwable t)
         {
             case die:
                 // policy to give 250ms grace time to
-                ScheduledExecutors.nonPeriodicTasks.schedule(() -> killer.killCurrentJVM(t), 250, TimeUnit.MILLISECONDS);
+                ScheduledExecutors.nonPeriodicTasks.schedule(() -> killer.killJVM(t), 250, TimeUnit.MILLISECONDS);
                 break;
             case die_immediate:
-                killer.killCurrentJVM(t);
+                killer.killJVM(t);
                 break;
             case ignore:
                 logger.error(t.getMessage());
                 break;
         }
     }
 
+    public static void registerShutdownHook(Thread hook, Runnable runOnHookRemoved)
+    {
+        Runtime.getRuntime().addShutdownHook(hook);
+        shutdownHooks.add(Pair.create(hook, runOnHookRemoved));
+    }
+
+    public static void removeShutdownHooks()
+    {
+        Throwable err = null;
+        for (Pair<Thread, Runnable> hook : shutdownHooks)
+        {
+            err = Throwables.perform(err,
+                                     () -> Runtime.getRuntime().removeShutdownHook(hook.left),
+                                     hook.right::run);
+        }
+
+        if (err != null)
+            logger.error("Got error when removing shutdown hook(s): {}", err.getMessage(), err);
+
+        shutdownHooks.clear();
```

Comment:
Is there a need for a special care if a hook wasn't removed due to error? Can such situation happen?

Reply:
tbh, I don't know what else could be done apart from logging the error (which it does).